File: android/src/main/kotlin/twilio/flutter/twilio_programmable_chat/methods/MessagesMethods.kt
package twilio.flutter.twilio_programmable_chat.methods

import com.twilio.chat.CallbackListener
import com.twilio.chat.Channel
import com.twilio.chat.ErrorInfo
import com.twilio.chat.Message
import com.twilio.chat.ProgressListener
import com.twilio.chat.StatusListener
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.FileInputStream
import twilio.flutter.twilio_programmable_chat.Mapper
import twilio.flutter.twilio_programmable_chat.TwilioProgrammableChatPlugin

object MessagesMethods {
    fun sendMessage(call: MethodCall, result: MethodChannel.Result) {
        val options = call.argument<Map<String, Any>>("options")
                ?: return result.error("ERROR", "Missing 'options'", null)
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        val messageOptions = Message.options()
        if (options["body"] != null) {
            messageOptions.withBody(options["body"] as String)
        }
        if (options["attributes"] != null) {
            messageOptions.withAttributes(Mapper.mapToAttributes(options["attributes"] as Map<String, Any>?))
        }
        if (options["input"] != null) {
            val input = options["input"] as String
            val mimeType = options["mimeType"] as String?
                    ?: return result.error("ERROR", "Missing 'mimeType' in MessageOptions", null)

            messageOptions.withMedia(FileInputStream(input), mimeType)
            if (options["filename"] != null) {
                messageOptions.withMediaFileName(options["filename"] as String)
            }

            if (options["mediaProgressListenerId"] != null) {
                messageOptions.withMediaProgressListener(object : ProgressListener() {
                    override fun onStarted() {
                        TwilioProgrammableChatPlugin.mediaProgresSink?.success({
                            "mediaProgressListenerId" to options["mediaProgressListenerId"]
                            "name" to "started"
                        })
                    }

                    override fun onProgress(bytes: Long) {
                        TwilioProgrammableChatPlugin.mediaProgresSink?.success({
                            "mediaProgressListenerId" to options["mediaProgressListenerId"]
                            "name" to "progress"
                            "data" to bytes
                        })
                    }

                    override fun onCompleted(mediaSid: String) {
                        TwilioProgrammableChatPlugin.mediaProgresSink?.success({
                            "mediaProgressListenerId" to options["mediaProgressListenerId"]
                            "name" to "completed"
                            "data" to mediaSid
                        })
                    }
                })
            }
        }

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.sendMessage (Channels.getChannel) => onSuccess")

                channel.messages.sendMessage(messageOptions, object : CallbackListener<Message>() {
                    override fun onSuccess(message: Message) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.sendMessage (Message.sendMessage) => onSuccess")
                        result.success(Mapper.messageToMap(message))
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.sendMessage (Message.sendMessage) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.sendMessage (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun removeMessage(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)
        val messageIndex = call.argument<Long>("messageIndex")
                ?: return result.error("ERROR", "Missing 'messageIndex'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Channels.getChannel) => onSuccess")

                channel.messages.getMessageByIndex(messageIndex, object : CallbackListener<Message>() {
                    override fun onSuccess(message: Message) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.getMessageByIndex) => onSuccess")

                        channel.messages.removeMessage(message, object : StatusListener() {
                            override fun onSuccess() {
                                TwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.removeMessage) => onSuccess")
                                result.success(null)
                            }

                            override fun onError(errorInfo: ErrorInfo) {
                                TwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.removeMessage) => onError: $errorInfo")
                                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                            }
                        })
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.getMessageByIndex) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.sendMessage (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getMessagesBefore(call: MethodCall, result: MethodChannel.Result) {
        val index = call.argument<Long>("index")
                ?: return result.error("ERROR", "Missing 'index'", null)
        val count = call.argument<Int>("count")
                ?: return result.error("ERROR", "Missing 'count'", null)
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesBefore (Channels.getChannel) => onSuccess")

                channel.messages.getMessagesBefore(index, count, object : CallbackListener<List<Message>>() {
                    override fun onSuccess(messages: List<Message>) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesBefore (Message.getMessagesBefore) => onSuccess")
                        val messagesListMap = messages?.map { Mapper.messageToMap(it) }
                        result.success(messagesListMap)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesBefore (Message.getMessagesBefore) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesBefore (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getMessagesAfter(call: MethodCall, result: MethodChannel.Result) {
        val index = call.argument<Int>("index")?.toLong()
                ?: return result.error("ERROR", "Missing 'index'", null)
        val count = call.argument<Int>("count")
                ?: return result.error("ERROR", "Missing 'count'", null)
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesAfter (Channels.getChannel) => onSuccess")

                channel.messages.getMessagesAfter(index, count, object : CallbackListener<List<Message>>() {
                    override fun onSuccess(messages: List<Message>) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesAfter (Message.getMessagesAfter) => onSuccess")
                        val messagesListMap = messages?.map { Mapper.messageToMap(it) }
                        result.success(messagesListMap)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesAfter (Message.getMessagesAfter) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesAfter (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getLastMessages(call: MethodCall, result: MethodChannel.Result) {
        val count = call.argument<Int>("count")
                ?: return result.error("ERROR", "Missing 'count'", null)
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getLastMessages (Channels.getChannel) => onSuccess")

                channel.messages.getLastMessages(count, object : CallbackListener<List<Message>>() {
                    override fun onSuccess(messages: List<Message>) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getLastMessages (Message.getLastMessages) => onSuccess")
                        val messagesListMap = messages.map { Mapper.messageToMap(it) }
                        result.success(messagesListMap)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getLastMessages (Message.getLastMessages) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getLastMessages (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getMessageByIndex(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)
        val messageIndex = call.argument<Long>("messageIndex")
                ?: return result.error("ERROR", "Missing 'messageIndex'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessageByIndex (Channels.getChannel) => onSuccess")

                channel.messages.getMessageByIndex(messageIndex, object : CallbackListener<Message>() {
                    override fun onSuccess(message: Message) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessageByIndex (Message.getMessageByIndex) => onSuccess")
                        result.success(Mapper.messageToMap(message))
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessageByIndex (Message.getMessageByIndex) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.getMessageByIndex (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun setLastConsumedMessageIndexWithResult(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)
        val lastConsumedMessageIndex = call.argument<Long>("lastConsumedMessageIndex")
                ?: return result.error("ERROR", "Missing 'lastConsumedMessageIndex'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.setLastConsumedMessageIndexWithResult (Channels.getChannel) => onSuccess")

                channel.messages.setLastConsumedMessageIndexWithResult(lastConsumedMessageIndex, object : CallbackListener<Long>() {
                    override fun onSuccess(newIndex: Long) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.setLastConsumedMessageIndexWithResult (Message.setLastConsumedMessageIndexWithResult) => onSuccess: $newIndex")
                        result.success(newIndex)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.setLastConsumedMessageIndexWithResult (Message.setLastConsumedMessageIndexWithResult) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.setLastConsumedMessageIndexWithResult (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun advanceLastConsumedMessageIndexWithResult(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)
        val lastConsumedMessageIndex = call.argument<Long>("lastConsumedMessageIndex")
                ?: return result.error("ERROR", "Missing 'lastConsumedMessageIndex'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.advanceLastConsumedMessageIndexWithResult (Channels.getChannel) => onSuccess")

                channel.messages.advanceLastConsumedMessageIndexWithResult(lastConsumedMessageIndex, object : CallbackListener<Long>() {
                    override fun onSuccess(newIndex: Long) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.advanceLastConsumedMessageIndexWithResult (Message.advanceLastConsumedMessageIndexWithResult) => onSuccess: $newIndex")
                        result.success(newIndex)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.advanceLastConsumedMessageIndexWithResult (Message.advanceLastConsumedMessageIndexWithResult) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.advanceLastConsumedMessageIndexWithResult (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun setAllMessagesConsumedWithResult(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.setAllMessagesConsumedWithResult (Channels.getChannel) => onSuccess")

                channel.messages.setAllMessagesConsumedWithResult(object : CallbackListener<Long>() {
                    override fun onSuccess(index: Long) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.setAllMessagesConsumedWithResult (Message.setAllMessagesConsumedWithResult) => onSuccess: $index")
                        result.success(index)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.setAllMessagesConsumedWithResult (Message.setAllMessagesConsumedWithResult) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.setAllMessagesConsumedWithResult (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun setNoMessagesConsumedWithResult(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.channels?.getChannel(channelSid, object : CallbackListener<Channel>() {
            override fun onSuccess(channel: Channel) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.setNoMessagesConsumedWithResult (Channels.getChannel) => onSuccess")

                channel.messages.setNoMessagesConsumedWithResult(object : CallbackListener<Long>() {
                    override fun onSuccess(index: Long) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.setNoMessagesConsumedWithResult (Message.setNoMessagesConsumedWithResult) => onSuccess: $index")
                        result.success(index)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        TwilioProgrammableChatPlugin.debug("MessagesMethods.setNoMessagesConsumedWithResult (Message.setNoMessagesConsumedWithResult) => onError: $errorInfo")
                        result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("MessagesMethods.setNoMessagesConsumedWithResult (Channels.getChannel) => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }
}

File: ios/Classes/Methods/MessagesMethods.swift
import Flutter
import Foundation
import TwilioChatClient

public class MessagesMethods {
    // swiftlint:disable:next cyclomatic_complexity
    public static func sendMessage(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let options = arguments["options"] as? [String: Any],
            let channelSid = arguments["channelSid"] as? String else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        let messageOptions = TCHMessageOptions(), mediaProgressListenerId = options["mediaProgressListenerId"]

        if let bodyRaw = options["body"], let body = bodyRaw as? String {
            messageOptions.withBody(body)
        }

        if let attributes = options["attributes"] as? [String: Any?] {
            messageOptions.withAttributes(Mapper.dictToAttributes(attributes)) { (result: TCHResult) in
                if !result.isSuccessful() {
                    flutterResult(FlutterError(code: "ERROR", message: "Failed to set attributes on message with error: \(String(describing: result.error))", details: nil))
                }
            }
        }

        if let input = options["input"] as? String {
            guard let mimeType = options["mimeType"] as? String else {
                return flutterResult(FlutterError(code: "ERROR", message: "Missing 'mimeType' in MessageOptions", details: nil))
            }

            if let inputStream = InputStream(fileAtPath: input) {
                messageOptions.withMediaStream(inputStream, contentType: mimeType, defaultFilename: options["filename"] as? String,
                                               onStarted: {
                                                if let id = mediaProgressListenerId, let sink = SwiftTwilioProgrammableChatPlugin.mediaProgressSink {
                                                    sink(["mediaProgressListenerId": id, "name": "started"])
                                                }},
                                               onProgress: { (bytes: UInt) in
                                                if let id = mediaProgressListenerId, let sink = SwiftTwilioProgrammableChatPlugin.mediaProgressSink {
                                                    sink(["mediaProgressListenerId": id, "name": "progress", "data": bytes])
                                                }},
                                               onCompleted: { (mediaSid: String) in
                                                if let id = mediaProgressListenerId, let sink = SwiftTwilioProgrammableChatPlugin.mediaProgressSink {
                                                    sink(["mediaProgressListenerId": id, "name": "completed", "data": mediaSid])
                                                }})
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving file for upload from `\(input)`", details: nil))
            }
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.sendMessage(with: messageOptions, completion: { (result: TCHResult, message: TCHMessage?) in
                    if result.isSuccessful(), let message = message {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.sendMessage (Message.sendMessage) => onSuccess")
                        flutterResult(Mapper.messageToDict(message, channelSid: channelSid))
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.sendMessage (Message.sendMessage) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error sending message with options `\(String(describing: messageOptions))`", details: nil))
                    }
                })
            }
        })
    }

    public static func removeMessage(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String,
            let messageIndex = arguments["messageIndex"] as? NSNumber else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.message(withIndex: messageIndex, completion: { (result: TCHResult, message: TCHMessage?) in
                    if result.isSuccessful(), let message = message {
                        channel.messages?.remove(message, completion: { (result: TCHResult) in
                            if result.isSuccessful() {
                                SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.removeMessage) => onSuccess")
                                flutterResult(nil)
                            } else {
                                SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.removeMessage) => onError: \(String(describing: result.error))")
                                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving message (index: \(channelSid)) from channel (sid: \(channelSid))", details: nil))
                            }
                        })
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.messageWithIndex) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving message (index: \(channelSid)) from channel (sid: \(channelSid))", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
            }
        })
    }

    public static func getMessagesBefore(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String,
            let index = arguments["index"] as? UInt,
            let count = arguments["count"] as? UInt else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.getBefore(index, withCount: count, completion: { (result: TCHResult, messages: [TCHMessage]?) in
                    if result.isSuccessful(), let messages = messages {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesBefore (Messages.getBefore) => onSuccess")
                        let messagesMap = messages.map { (message: TCHMessage) -> [String: Any?] in
                            return Mapper.messageToDict(message, channelSid: channelSid)
                        }
                        flutterResult(messagesMap)
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesBefore (Messages.getBefore) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving \(count) messages before message (index: \(index)) from channel (sid: \(channelSid))", details: nil))
                    }
                })
            }
        })
    }

    public static func getMessagesAfter(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String,
            let index = arguments["index"] as? UInt,
            let count = arguments["count"] as? UInt else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.getAfter(index, withCount: count, completion: { (result: TCHResult, messages: [TCHMessage]?) in
                    if result.isSuccessful(), let messages = messages {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesAfter (Messages.getAfter) => onSuccess")
                        let messagesMap = messages.map { (message: TCHMessage) -> [String: Any?] in
                            return Mapper.messageToDict(message, channelSid: channelSid)
                        }
                        flutterResult(messagesMap)
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.getMessagesAfter (Messages.getAfter) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving \(count) messages after message (index: \(index)) from channel (sid: \(channelSid))", details: nil))
                    }
                })
            }
        })
    }

    public static func getLastMessages(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let count = arguments["count"] as? UInt,
            let channelSid = arguments["channelSid"] as? String else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.getLastWithCount(count, completion: { (result: TCHResult, messages: [TCHMessage]?) in
                    if result.isSuccessful(), let messages = messages {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.getLastMessages (Messages.getLast) => onSuccess")
                        let messagesMap = messages.map({ (message: TCHMessage) -> [String: Any?] in
                            return Mapper.messageToDict(message, channelSid: channelSid)
                        })
                        flutterResult(messagesMap)
                    } else {
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving last \(count) messages for channel with sid '\(channelSid)'", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
            }
        })
    }

    public static func getMessageByIndex(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String,
            let messageIndex = arguments["messageIndex"] as? NSNumber else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.message(withIndex: messageIndex, completion: { (result: TCHResult, message: TCHMessage?) in
                    if result.isSuccessful(), let message = message {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.getMessageByIndex (Messages.messageWithIndex) => onSuccess")
                        flutterResult(Mapper.messageToDict(message, channelSid: channelSid))
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.removeMessage (Messages.messageWithIndex) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving message (index: \(channelSid)) from channel (sid: \(channelSid))", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
            }
        })
    }

    public static func setLastConsumedMessageIndexWithResult(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String,
            let lastConsumedMessageIndex = arguments["lastConsumedMessageIndex"] as? NSNumber else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel, let messages = channel.messages {
                messages.setLastConsumedMessageIndex(lastConsumedMessageIndex, completion: { (result: TCHResult, count: UInt) in
                    if result.isSuccessful() {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.setLastConsumedMessageIndexWithResult (Messages.setLastConsumedMessageIndex) => onSuccess")
                        flutterResult(count)
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.setLastConsumedMessageIndexWithResult (Messages.setLastConsumedMessageIndex) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error setting last consumed message index (index: \(lastConsumedMessageIndex)) for channel (sid: \(channelSid))", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
            }
        })
    }

    public static func advanceLastConsumedMessageIndexWithResult(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String,
            let lastConsumedMessageIndex = arguments["lastConsumedMessageIndex"] as? NSNumber else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.advanceLastConsumedMessageIndex(lastConsumedMessageIndex, completion: { (result: TCHResult, count: UInt) in
                    if result.isSuccessful() {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.advanceLastConsumedMessageIndexWithResult (Messages.advanceLastConsumedMessageIndex) => onSuccess")
                        flutterResult(count)
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.advanceLastConsumedMessageIndexWithResult (Messages.advanceLastConsumedMessageIndex) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error advancing last consumed message index (index: \(lastConsumedMessageIndex)) for channel with sid '\(channelSid)'", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
            }
        })
    }

    public static func setAllMessagesConsumedWithResult(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel {
                channel.messages?.setAllMessagesConsumedWithCompletion({ (result: TCHResult, count: UInt) in
                    if result.isSuccessful() {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.setAllMessagesConsumedWithResult (Messages.setAllMessagesConsumed) => onSuccess")
                        flutterResult(count)
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.setAllMessagesConsumedWithResult (Messages.setAllMessagesConsumed) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error setting all messages consumed for channel (sid: \(channelSid))", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
            }
        })
    }

    public static func setNoMessagesConsumedWithResult(_ call: FlutterMethodCall, result flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
            let channelSid = arguments["channelSid"] as? String else {
                return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channel = channel, let messages = channel.messages {
                messages.setNoMessagesConsumedWithCompletion({ (result: TCHResult, count: UInt) in
                    if result.isSuccessful() {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.setNoMessagesConsumedWithResult (Messages.setNoMessagesConsumed) => onSuccess")
                        flutterResult(count)
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("MessagesMethods.setNoMessagesConsumedWithResult (Messages.setNoMessagesConsumed) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error setting no messages consumed for channel (sid: \(channelSid))", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel with sid '\(channelSid)'", details: nil))
            }
        })
    }
}
